flaky test_no_reconnect[--no-nanny] TimeoutError: Operation timed out after None seconds

failure on the last main build: https://github.com/dask/distributed/runs/4888280793?check_suite_focus=true#step:12:1358

```
=================================== FAILURES ===================================
________________________ test_no_reconnect[--no-nanny] _________________________

nanny = '--no-nanny'
loop = <tornado.platform.asyncio.AsyncIOLoop object at 0x7fbee0de8400>

    @pytest.mark.slow
    @pytest.mark.parametrize("nanny", ["--nanny", "--no-nanny"])
    def test_no_reconnect(nanny, loop):
        with popen(["dask-scheduler", "--no-dashboard"]) as sched:
>           wait_for_port(("127.0.0.1", 8786))

distributed/cli/tests/test_dask_worker.py:154: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

address = ('127.0.0.1', 8786), timeout = -0.08133506774902344

    def wait_for_port(address, timeout=5):
        assert isinstance(address, tuple)
        deadline = time() + timeout
    
        while True:
            timeout = deadline - time()
            if timeout < 0:
>               raise RuntimeError(f"Failed to connect to {address}")
E               RuntimeError: Failed to connect to ('127.0.0.1', 8786)

distributed/utils_test.py:1185: RuntimeError
----------------------------- Captured stdout call -----------------------------


Print from stderr
  /
=================

distributed.scheduler - INFO - -----------------------------------------------
distributed.http.proxy - INFO - To route to workers diagnostics web server please install jupyter-server-proxy: python -m pip install jupyter-server-proxy
distributed.scheduler - INFO - -----------------------------------------------
distributed.scheduler - INFO - Clear task state
distributed.scheduler - INFO -   Scheduler at:  tcp://10.212.11.235:8786
distributed.scheduler - INFO -   dashboard at:                     :8787
distributed.scheduler - INFO - End scheduler at 'tcp://10.212.11.235:8786'
Traceback (most recent call last):
  File "/Users/runner/miniconda3/envs/dask-distributed/bin/dask-scheduler", line 33, in <module>
    sys.exit(load_entry_point('distributed', 'console_scripts', 'dask-scheduler')())
  File "/Users/runner/work/distributed/distributed/distributed/cli/dask_scheduler.py", line 217, in go
    main()
  File "/Users/runner/miniconda3/envs/dask-distributed/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/runner/miniconda3/envs/dask-distributed/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/runner/miniconda3/envs/dask-distributed/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/runner/miniconda3/envs/dask-distributed/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/runner/work/distributed/distributed/distributed/cli/dask_scheduler.py", line 208, in main
    loop.run_sync(run)
  File "/Users/runner/miniconda3/envs/dask-distributed/lib/python3.8/site-packages/tornado/ioloop.py", line 529, in run_sync
    raise TimeoutError("Operation timed out after %s seconds" % timeout)
tornado.util.TimeoutError: Operation timed out after None seconds
```


the `TimeoutError: Operation timed out after None seconds` is due to the scheduler task being cancelled or propagating a `CancelledError` incorrectly even though `run_sync` has no timeout.

--- a/distributed/cli/tests/test_dask_worker.py
+++ b/distributed/cli/tests/test_dask_worker.py
@@ -1,4 +1,7 @@
 import asyncio
+import contextvars
+import functools
+import sys
 
 import pytest
 from click.testing import CliRunner
@@ -19,13 +22,28 @@ from distributed.compatibility import LINUX
 from distributed.deploy.utils import nprocesses_nthreads
 from distributed.metrics import time
 from distributed.utils import parse_ports, sync
-from distributed.utils_test import (
-    gen_cluster,
-    popen,
-    requires_ipv6,
-    terminate_process,
-    wait_for_port,
-)
+from distributed.utils_test import gen_cluster, popen, requires_ipv6
+
+if sys.version_info >= (3, 9):
+    from asyncio import to_thread
+else:
+
+    async def to_thread(*func_args, **kwargs):
+        """Asynchronously run function *func* in a separate thread.
+        Any *args and **kwargs supplied for this function are directly passed
+        to *func*. Also, the current :class:`contextvars.Context` is propagated,
+        allowing context variables from the main thread to be accessed in the
+        separate thread.
+        Return a coroutine that can be awaited to get the eventual result of *func*.
+
+        backport from
+        https://github.com/python/cpython/blob/3f1ea163ea54513e00e0e9d5442fee1b639825cc/Lib/asyncio/threads.py#L12-L25
+        """
+        func, *args = func_args
+        loop = asyncio.get_running_loop()
+        ctx = contextvars.copy_context()
+        func_call = functools.partial(ctx.run, func, *args, **kwargs)
+        return await loop.run_in_executor(None, func_call)
 
 
 def test_nanny_worker_ports(loop):
@@ -149,24 +167,55 @@ def test_no_nanny(loop):
 
 @pytest.mark.slow
 @pytest.mark.parametrize("nanny", ["--nanny", "--no-nanny"])
-def test_no_reconnect(nanny, loop):
-    with popen(["dask-scheduler", "--no-dashboard"]) as sched:
-        wait_for_port(("127.0.0.1", 8786))
+@pytest.mark.asyncio
+async def test_no_reconnect(nanny):
+    async with Scheduler(dashboard_address=":0") as s, Client(
+        s.address, asynchronous=True
+    ) as c:
         with popen(
             [
                 "dask-worker",
-                "tcp://127.0.0.1:8786",
+                s.address,
                 "--no-reconnect",
                 nanny,
                 "--no-dashboard",
             ]
         ) as worker:
-            sleep(2)
-            terminate_process(sched)
-        start = time()
-        while worker.poll() is None:
-            sleep(0.1)
-            assert time() < start + 30
+            await c.wait_for_workers(1)
+
+            (comm,) = s.stream_comms.values()
+            comm.abort()
+
+            await to_thread(worker.communicate, timeout=30)
+            assert worker.returncode == 0
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize("nanny", ["--nanny", "--no-nanny"])
+@pytest.mark.asyncio
+async def test_reconnect(nanny):
+    async with Scheduler(dashboard_address=":0") as s, Client(
+        s.address, asynchronous=True
+    ) as c:
+        with popen(
+            [
+                "dask-worker",
+                s.address,
+                "--reconnect",
+                nanny,
+                "--no-dashboard",
+            ]
+        ) as worker:
+            await c.wait_for_workers(1)
+
+            (comm,) = s.stream_comms.values()
+            comm.abort()
+
+            await c.wait_for_workers(1)
+            await s.close()
+
+            await to_thread(worker.communicate, timeout=30)
+            assert worker.returncode == 0
 
 
 def test_resources(loop):

--- a/distributed/cli/tests/test_dask_worker.py
+++ b/distributed/cli/tests/test_dask_worker.py
@@ -181,12 +181,14 @@ async def test_no_reconnect(nanny):
                 "--no-dashboard",
             ]
         ) as worker:
-            await c.wait_for_workers(1)
+            # roundtrip works
+            assert await c.submit(lambda x: x + 1, 10) == 11
 
             (comm,) = s.stream_comms.values()
             comm.abort()
 
-            await to_thread(worker.communicate, timeout=30)
+            # worker terminates as soon as the connection is aborted
+            await to_thread(worker.communicate, timeout=5)
             assert worker.returncode == 0
 
 
@@ -206,15 +208,18 @@ async def test_reconnect(nanny):
                 "--no-dashboard",
             ]
         ) as worker:
-            await c.wait_for_workers(1)
+            # roundtrip works
+            await c.submit(lambda x: x + 1, 10) == 11
 
             (comm,) = s.stream_comms.values()
             comm.abort()
 
-            await c.wait_for_workers(1)
-            await s.close()
+            # roundtrip still works, which means the worker reconnected
+            assert await c.submit(lambda x: x + 1, 11) == 12
 
-            await to_thread(worker.communicate, timeout=30)
+            # closing the scheduler cleanly does terminate the worker
+            await s.close()
+            await to_thread(worker.communicate, timeout=5)
             assert worker.returncode == 0
 
 

--- a/distributed/tests/test_utils_test.py
+++ b/distributed/tests/test_utils_test.py
@@ -15,7 +15,6 @@ import dask.config
 from distributed import Client, Nanny, Scheduler, Worker, config, default_client
 from distributed.core import Server, rpc
 from distributed.metrics import time
-from distributed.utils import get_ip
 from distributed.utils_test import (
     _LockedCommPool,
     _UnhashableCallable,
@@ -27,7 +26,6 @@ from distributed.utils_test import (
     inc,
     new_config,
     tls_only_security,
-    wait_for_port,
 )
 
 
@@ -230,26 +228,6 @@ def _listen(delay=0):
         yield serv
     finally:
         t.join(5.0)
-
-
-def test_wait_for_port():
-    t1 = time()
-    with pytest.raises(RuntimeError):
-        wait_for_port((get_ip(), 9999), 0.5)
-    t2 = time()
-    assert t2 - t1 >= 0.5
-
-    with _listen(0) as s1:
-        t1 = time()
-        wait_for_port(s1.getsockname())
-        t2 = time()
-        assert t2 - t1 <= 1.0
-
-    with _listen(1) as s1:
-        t1 = time()
-        wait_for_port(s1.getsockname())
-        t2 = time()
-        assert t2 - t1 <= 2.0
 
 
 def test_new_config():

--- a/distributed/utils_test.py
+++ b/distributed/utils_test.py
@@ -1198,23 +1198,6 @@ def popen(args, **kwargs):
                 print(out.decode())
 
 
-def wait_for_port(address, timeout=5):
-    assert isinstance(address, tuple)
-    deadline = time() + timeout
-
-    while True:
-        timeout = deadline - time()
-        if timeout < 0:
-            raise RuntimeError(f"Failed to connect to {address}")
-        try:
-            sock = socket.create_connection(address, timeout=timeout)
-        except OSError:
-            pass
-        else:
-            sock.close()
-            break
-
-
 def wait_for(predicate, timeout, fail_func=None, period=0.001):
     deadline = time() + timeout
     while not predicate():

--- a/distributed/utils_test.py
+++ b/distributed/utils_test.py
@@ -1147,7 +1147,7 @@ def raises(func, exc=Exception):
         return True
 
 
-def terminate_process(proc):
+def _terminate_process(proc):
     if proc.poll() is None:
         if sys.platform.startswith("win"):
             proc.send_signal(signal.CTRL_BREAK_EVENT)
@@ -1186,7 +1186,7 @@ def popen(args, **kwargs):
 
     finally:
         try:
-            terminate_process(proc)
+            _terminate_process(proc)
         finally:
             # XXX Also dump stdout if return code != 0 ?
             out, err = proc.communicate()

